### DISCUSSION TITLE
How to use SPI with gokrazy?

### DISCUSSION THREAD

**Discussion by x86ed** [gokrazy/gokrazy#224] [Q&A]
I want to enable SPI, UART and I2C on my raspberry pi but it's not clear what commands/config I need to do to enable it on my board through gok

**Comment by stapelberg** [comment]
Hey @x86ed! 

If you want to make changes to the kernel, use a modified version of https://github.com/gokrazy/kernel/, like you would swap out any other Go package: https://gokrazy.org/development/modules/#building-local-code-the-replace-directive

The kernel build we use stores its kernel config in cmd/gokr-build-kernel/build.go. It might be more convenient to use `ARCH=arm64 make menuconfig` in a separate checkout of the Linux kernel to figure out which config options you need/want to change.

In general, we accept changes to the kernel repository, as long as they don’t have adverse affects for other users (e.g. require hard-to-maintain patches, or disable other features that people rely on). UART is already enabled, and hopefully SPI and I2C should be quick and uncontroversial, I think…?

Hope this helps

**Reply by damdo** [reply]
Ahh it looks like you beat me to that @stapelberg :)

**Comment by damdo** [comment]
Hey, those configs must be set in the [Linux kernel configuration we define for gokrazy](https://github.com/gokrazy/kernel), unless they are already.

Here, for example, we have the configuration to enable the I2C and the SPI support in the kernel:
https://github.com/gokrazy/kernel/blob/b015f4d0ae408c30b52679aa27399e281ed7c149/cmd/gokr-build-kernel/build.go#L23-L25
The recommended way to interact with those buses is to use the [periph.io](https://periph.io/project/library/) libs.

To my knowledge, the UART serial console is also enabled (if that's what you are looking for), see: https://github.com/gokrazy/tools/blob/231dea4b57042ff090c0a94fa03abe1dab9e3a3f/internal/packer/write.go#L76
You can tweak its config with: https://gokrazy.org/userguide/instance-config/#serialconsole

**Reply by x86ed** [reply]
Is it possible to enable I2C in the same manner as the serial console?

**Reply by damdo** [reply]
As I alluded to, I2C is already enabled in the default gokrazy kernel.
So you can already use it in your Go programs when they run on gokrazy.
Try: https://pkg.go.dev/periph.io/x/periph/conn/i2c/i2creg#example-package

**Reply by x86ed** [reply]
apologies the version above doesn't seem compatible with the current version of periph which appears to be v3.8.2

**Reply by damdo** [reply]
I'm using periph and gokrazy successfully for GPIO.
I've never tried the I2C bus, but on paper everything _should_ be compatible.

What errors are you seeing?

**Reply by x86ed** [reply]
```
2023/10/04 23:16:44 sysfs-i2c: remote I/O error
2023/10/04 23:16:44 gokrazy: exit status 1
```

is what i get when trying to run the sample you provided above using 
```
	periph.io/x/host/v3 v3.8.2
	periph.io/x/conn/v3 v3.7.0 // indirect
	periph.io/x/devices/v3 v3.7.1
```
validated that the gpio code found [here](https://gokrazy.org/development/gpio/) works with the libraries listed above even though those aren't used in the existing code.
you can see the LED triggered from code here:
![FF5BF4D2-72E6-4AFB-AD80-E7B89870B4E4_1_102_o](https://github.com/gokrazy/gokrazy/assets/660891/e65540d5-ae02-47c6-bb04-fe4d84cb94d1)


I've also tried this code against an eInk display [waveshare 2.13in](https://www.waveshare.com/wiki/2.13inch_e-Paper_HAT_Manual#Working_With_Raspberry_Pi)

using this code:
```go
package main

import (
	"image"
	"log"

	"github.com/fogleman/gg"
	"github.com/golang/freetype/truetype"
	"golang.org/x/image/font"
	"golang.org/x/image/font/basicfont"
	"golang.org/x/image/font/gofont/goregular"
	"golang.org/x/image/math/fixed"
	"periph.io/x/devices/v3/epd"

	"periph.io/x/conn/v3/spi/spireg"

	"periph.io/x/devices/v3/ssd1306/image1bit"
	"periph.io/x/host/v3"
)

func main() {
	// Make sure periph is initialized.
	if _, err := host.Init(); err != nil {
		log.Fatal(err)
	}

	// Use spireg SPI bus registry to find the first available SPI bus.
	b, err := spireg.Open("")
	if err != nil {
		log.Fatal(err)
	}
	defer b.Close()

	dev, err := epd.NewSPIHat(b, &epd.EPD2in13) // Display config and size
	if err != nil {
		log.Fatalf("failed to initialize epd: %v", err)
	}

	// Draw on it.
	img := image1bit.NewVerticalLSB(dev.Bounds())

	f := basicfont.Face7x13
	drawer := font.Drawer{
		Dst:  img,
		Src:  &image.Uniform{image1bit.On},
		Face: f,
		Dot:  fixed.P(0, img.Bounds().Dy()-1-f.Descent),
	}
	drawer.DrawString("Hello from periph!")

	if err := dev.Draw(dev.Bounds(), img, image.Point{}); err != nil {
		log.Fatal(err)
	}
	dev.DisplayFrame() // After drawing on the display, you have to show the frame
}
```

and this is producing the error: 
`spireg: no port found; did you forget to call Init()?`

**Reply by x86ed** [reply]
Oops I'm an idiot I'm looking for SPI not i2c

**Reply by stapelberg** [reply]
Is this code known working on a Raspbian installation? If no, try that first. If yes, does it work on Debian or Fedora (not Raspbian) as well?

If yes, can you compare the `strace` output between Raspbian, Debian/Fedora and gokrazy? I’m wondering which driver the respective OSes use. Raspbian might have a custom kernel driver that works differently from upstream Linux’s, and what’s enabled in gokrazy might be yet another subset of that.

**Reply by x86ed** [reply]
https://github.com/gokrazy/gokrazy/issues/47 I found this old PR that I believe should fix the issue but, the PR doesn't provide the code or kernel configuration the author used to test the SPI port. I assume they used periph.io given the stack trace in the thread they posted.


Typically a program called `raspi-config` [repo](https://github.com/RPi-Distro/raspi-config) is used to set the kernel configuration in raspbian. I've used this in the past with go/periph on raspbian to connect to SPI ports. My biggest hunch is that  I'm missing a configuration task on the kernel setup that should enable it. But I'm really new to altering Kernel configuration so I'm kinda unsure what i should be looking for with strace and how to run it.

**Reply by stapelberg** [reply]
The pull request you linked to was merged years ago, so I was confused why you were still running into issues.

But now that you mention raspi-config, I found https://www.raspberrypi-spy.co.uk/2014/08/enabling-the-spi-interface-on-the-raspberry-pi/, which says that `config.txt` (the early boot firmware configuration) should contain `dtparam=spi=on`.

We currently don’t offer a way to customize what goes into config.txt in gokrazy. However, you can clone the gokrazy kernel repository  and manually edit config.txt (see also https://gokrazy.org/development/modules/#building-local-code-the-replace-directive). Try making that change and see if that helps.

**Reply by stapelberg** [reply]
By the way, we now do have a way to customize config.txt: https://gokrazy.org/userguide/instance-config/#bootloader

**Comment by x86ed** [comment]
I cloned the kernel, added it to my gokrazy install and added the following lines to the `config.txt`
```
dtparam=spi=on

dtoverlay=spi1-1cs #enables SPI1 with 1 chip select
dtoverlay=spi1-2cs #enables SPI1 with 2 chip selects
dtoverlay=spi1-3cs #enables SPI1 with 3 chip selects
```
ran breakglass and did a `ls /dev`
got
```
autofs           loop1            ptyp0            serial0          tty20            tty37            tty53            ttyp1            vcs1
bus              loop2            ptyp1            shm              tty21            tty38            tty54            ttyp2            vcs2
cec0             loop3            ptyp2            snapshot         tty22            tty39            tty55            ttyp3            vcsa
console          loop4            ptyp3            snd              tty23            tty4             tty56            ttyp4            vcsa1
cpu_dma_latency  loop5            ptyp4            tty              tty24            tty40            tty57            ttyp5            vcsa2
dri              loop6            ptyp5            tty0             tty25            tty41            tty58            ttyp6            vcsu
fb0              loop7            ptyp6            tty1             tty26            tty42            tty59            ttyp7            vcsu1
full             mem              ptyp7            tty10            tty27            tty43            tty6             ttyp8            vcsu2
fuse             mmcblk0          ptyp8            tty11            tty28            tty44            tty60            ttyp9            vfio
gpiochip0        mmcblk0p1        ptyp9            tty12            tty29            tty45            tty61            ttypa            vga_arbiter
hwrng            mmcblk0p2        ptypa            tty13            tty3             tty46            tty62            ttypb            watchdog
i2c-0            mmcblk0p3        ptypb            tty14            tty30            tty47            tty63            ttypc            watchdog0
i2c-1            mmcblk0p4        ptypc            tty15            tty31            tty48            tty7             ttypd            zero
i2c-2            net              ptypd            tty16            tty32            tty49            tty8             ttype
kmsg             null             ptype            tty17            tty33            tty5             tty9             ttypf
kvm              port             ptypf            tty18            tty34            tty50            ttyAMA0          uinput
loop-control     ptmx             random           tty19            tty35            tty51            ttyS0            urandom
loop0            pts              rfkill           tty2             tty36            tty52            ttyp0            vcs
```

lsmod yields:
```
brcmfmac_wcc 16384 0 - Live 0xffffd5b1dd27e000
brcmfmac 266240 1 brcmfmac_wcc, Live 0xffffd5b1dd239000
brcmutil 16384 1 brcmfmac, Live 0xffffd5b1dd232000
```
there don't appear to be any spi entries. anything i can do to troubleshoot that?

**Reply by stapelberg** [reply]
> I cloned the kernel, added it to my gokrazy install and added the following lines to the `config.txt`

But did you also place the spi1-1cs.dtbo (and the other variants) from https://github.com/raspberrypi/firmware/tree/master/boot/overlays in the overlays/ directory in the kernel working directory? It should end up on the SD card on the first partition. You’ll need a recent build of the “gok” tool for that which includes commit https://github.com/gokrazy/tools/commit/c6f9ca1dab8cb30067d1d1c47242fcedd130e8de

**Reply by x86ed** [reply]
Nope haven't tried that. Is updating gok likely to break the existing flash i have?

**Reply by stapelberg** [reply]
No, it’s very unlikely that the recent gok changes will break your setup.

**Reply by x86ed** [reply]
here? 
<img width="312" alt="Screenshot 2023-10-21 at 3 32 52 PM" src="https://github.com/gokrazy/gokrazy/assets/660891/bcd62719-d2ca-4770-b6e3-299802a59777">

**Reply by stapelberg** [reply]
No, that’s the kernel modules directory. You’ll need to create a new “overlays” directory in your ~/Workspace/git/kernel first.

**Reply by x86ed** [reply]
is there a way of loading the dtoverlay command with gokrazy?  I've created the overlays but still haven't seen the devices  mount when i check with busybox

**Reply by stapelberg** [reply]
I don’t know what exactly you mean with “loading the dtoverlay command with gokrazy”, but I’ll say no. We currently don’t have a way to influence config.txt in gokrazy.

**Reply by damdo** [reply]
Hey @x86ed were you able to make SPI work in the end?
I'm interested in using it to pilot an SPI display in gokrazy, which I think it is similar to what you were attempting.

**Reply by stapelberg** [reply]
By the way, we now do have a way to customize config.txt: https://gokrazy.org/userguide/instance-config/#bootloader

**Reply by x86ed** [reply]
@damdo I didn't but I do want to explore the above. The issue I ran into was trying to replicate the bios settings in go crazy without having `raspi-config` just ended up spnning my wheels as I'm a bit out of my depth.

**Reply by x86ed** [reply]
@stapelberg will try to attempt using this later, to see if i can unblock myself

**Reply by Crocmagnon** [reply]
Hi, I was able to enable spi using https://gokrazy.org/userguide/instance-config/#bootloader 👍🏻 
At least it's detected by periph.io's `host.Init()` :)
